File: read_basic_ini_file.pl
#!/usr/bin/perl

# Program: read_basic_ini_file.pl
# Purpose: Basic Windows-style .ini file parser (without sections) sample
# Env: Perl5
# Author: James Briggs, 2019
# Licence: MIT
# Date: 2019 08 05
# Note: non-trivial example of file handling and parsing for you to customize

use strict;
use diagnostics;

use String::Util qw/trim/;
use Data::Dumper;

   my $DEBUG = 0;

   my $filepath = "config.ini";
   my $h = read_basic_ini_file($filepath);

   print Dumper($h) if $DEBUG; # show nested hash data structure

   # Do not use each:
   # http://blogs.perl.org/users/rurban/2014/04/do-not-use-each.html
   for my $k (sort keys %{$h}) {
      no warnings 'uninitialized';
      print "$k=$h->{$k}\n"; # use arrow notation to dereference hashref
   }

sub read_basic_ini_file {
   my ($filepath) = @_;

   open X, "<$filepath" or die "error: $!";

   my %h; # hash

   while (<X>) {
      print if $DEBUG;
      chomp;               # remove trailing newline for this OS

      next if /^$/ or      # skip blank line
              /^;|#/ or      # skip line starting with a comment character
              /^\[[^\]]*\]$/;  # skip section name like [section]

      my ($key, $value) = split(/=/, $_, 2);

      $h{trim($key)} = trim($value);
      print "$key=$value\n" if $DEBUG;
   }

   close X;

   return \%h; # reference to hash
}

File: read_basic_ini_file.rb
#!/usr/bin/ruby
#
# Purpose: Basic Windows-style .ini file parser (without sections) sample
# Env: Ruby
# Author: James Briggs, 2019
# Licence: MIT
# Date: 2019 08 05
# Note: non-trivial example of file handling and parsing for you to customize

def read_basic_ini_file(filepath)
   h = {}
   f = File.open(filepath, "r")
   f.each_line do |line|
      line = line.chomp
      if line =~ /^$/ or
         line =~ /^;|#/ or
         line =~ /^\[[^\]]*\]$/
         next
      end
      key, value = line.split('=', 2).map(&:strip)
      h[key] = value
   end
   f.close

   return h
end

h = read_basic_ini_file("config.ini")

h.keys.sort.each {|key| puts "#{key}=#{h[key]}" }

